corgi/tasks/errata_tool.py

```diff
@@ -35,15 +41,41 @@ def save_variant_cdn_repo_mapping() -> None:
     ErrataTool().save_variant_cdn_repo_mapping()
 
 
-@app.task(base=Singleton, autoretry_for=RETRYABLE_ERRORS, retry_kwargs=RETRY_KWARGS)
-def slow_save_errata_product_taxonomy(erratum_id: int) -> None:
-    logger.info(f"slow_save_errata_product_taxonomy called for {erratum_id}")
-    relation_builds = _get_relation_builds(erratum_id)
-    for build_id, build_type in relation_builds:
-        logger.info("Saving product taxonomy for build (%s, %s)", build_id, build_type)
-        # once all build's components are ingested we must save product taxonomy
-        sb = SoftwareBuild.objects.get(build_id=build_id, build_type=build_type)
-        sb.save_product_taxonomy()
+def associate_variant_with_build_streams(
+    build_id: str, build_type: str, variant_names: list[str]
+) -> bool:
+    # The SoftwareBuild should always exist, let the DoesNotExist error propagate if it doesn't
+    software_build = SoftwareBuild.objects.get(build_id=build_id, build_type=build_type)
+    build_streams = ProductStream.objects.filter(
+        components__software_build=software_build, active=True
+    )
+    any_created = False
+    for variant_name in variant_names:
+        # The CollectorErrataProductVariant should always exist, let the DoesNotExist error
+        # propagate if it doesn't
+        errata_variant = CollectorErrataProductVariant.objects.get(name=variant_name)
+        stream = build_streams.first()
+        # Only associate this variant to a single stream to keep one-to-many cardinality between
+        # ProductStream and ProductVariant
+        if stream and len(build_streams) == 1:
```

Comment:
We're going to repeatedly make DB calls for the stream inside the for-loop, but they're the same for each variant name. Let's move them out of the loop instead. We can also avoid looking up the variant when the stream condition is false / we don't do any linking:
```suggestion
    # This evaluates the build_streams queryset once and avoids further DB calls
    build_streams_count = len(build_streams)
    stream = None
    if build_streams_count:
        # Reuse the evaiuated queryset instead of calling .first() / doing a new DB lookup
        stream = build_streams[0]
    for variant_name in variant_names:
        # Only associate this variant to a single stream to keep one-to-many cardinality between
        # ProductStream and ProductVariant
        if stream and build_streams_count == 1:
            # The CollectorErrataProductVariant should always exist, let the DoesNotExist error
            # propagate if it doesn't
            errata_variant = CollectorErrataProductVariant.objects.get(name=variant_name)
```